File: packages/sdk/sdk-php/src/Workflows.php
<?php
declare(strict_types=1);

namespace Phaseo\Sdk;

use IteratorAggregate;
use RuntimeException;
use Traversable;

final class PageIterator implements IteratorAggregate
{
    public function __construct(
        private $fetchPage,
        private int $limit = 50,
        private int $offset = 0,
    ) {
        $this->limit = max(1, $this->limit);
        $this->offset = max(0, $this->offset);
    }

    public function getIterator(): Traversable
    {
        $offset = $this->offset;
        while (true) {
            $page = ($this->fetchPage)($this->limit, $offset);
            yield $page;
            $items = is_array($page['data'] ?? null) ? $page['data'] : [];
            if (!($page['has_more'] ?? false) || count($items) === 0) return;
            $offset += count($items);
        }
    }
}

final class JobHandle
{
    public function __construct(
        public readonly string $kind,
        public readonly string $id,
        private $fetch,
        private $status,
    ) {
        if (trim($id) === '') throw new \InvalidArgumentException('Job ID is required');
    }

    public function refresh(): array
    {
        return ($this->fetch)($this->id);
    }

    public function wait(float $intervalSeconds = 1.0, ?float $timeoutSeconds = null): array
    {
        $deadline = $timeoutSeconds === null ? null : microtime(true) + $timeoutSeconds;
        while (true) {
            $value = $this->refresh();
            $status = strtolower((string) ($this->status)($value));
            if ($status === 'completed') return $value;
            if (in_array($status, ['failed', 'cancelled', 'canceled', 'expired'], true)) {
                throw new RuntimeException("{$this->kind} job {$this->id} ended with status {$status}");
            }
            if ($deadline !== null && microtime(true) >= $deadline) throw new RuntimeException('Timed out waiting for Phaseo job');
            usleep((int) (max(0.001, $intervalSeconds) * 1_000_000));
        }
    }
}

File: packages/sdk/sdk-ruby/lib/phaseo_sdk/workflows.rb
module PhaseoSdk
  class PageEnumerator
    include Enumerable

    def initialize(limit: 50, offset: 0, &fetch_page)
      @limit = [limit.to_i, 1].max
      @offset = [offset.to_i, 0].max
      @fetch_page = fetch_page
    end

    def each
      return enum_for(:each) unless block_given?
      offset = @offset
      loop do
        page = @fetch_page.call(limit: @limit, offset: offset)
        yield page
        items = page[:data] || page["data"] || []
        has_more = page[:has_more] || page["has_more"]
        break unless has_more && !items.empty?
        offset += items.length
      end
    end
  end

  class JobHandle
    TERMINAL_FAILURES = %w[failed cancelled canceled expired].freeze

    attr_reader :kind, :id

    def initialize(kind:, id:, fetch:, status:)
      raise ArgumentError, "Job ID is required" if id.to_s.strip.empty?
      @kind, @id, @fetch, @status = kind, id, fetch, status
    end

    def refresh = @fetch.call(@id)

    def wait(interval: 1.0, timeout: nil)
      deadline = timeout && Process.clock_gettime(Process::CLOCK_MONOTONIC) + timeout
      loop do
        value = refresh
        current = @status.call(value).to_s.downcase
        return value if current == "completed"
        raise "#{kind} job #{id} ended with status #{current}" if TERMINAL_FAILURES.include?(current)
        raise "Timed out waiting for Phaseo job" if deadline && Process.clock_gettime(Process::CLOCK_MONOTONIC) >= deadline
        sleep([interval, 0.001].max)
      end
    end
  end
end
